src/v/serde/serde.h

```diff
@@ -166,9 +177,64 @@ inline constexpr auto const are_bytes_and_string_different = !(
 template<typename T>
 void write(iobuf&, T);
 
+template<class R, class P>
+int64_t checked_duration_cast_to_nanoseconds(
+  const std::chrono::duration<R, P>& duration) {
+    static_assert(
+      __has_builtin(__builtin_mul_overflow),
+      "__builtin_mul_overflow not supported.");
+    using nano_period = std::chrono::nanoseconds::period;
+    using nano_rep = typename std::chrono::nanoseconds::rep;
+    // This is how duration_cast determines the output type.
+    using output_type = typename std::common_type<R, nano_rep, intmax_t>::type;
+    using ratio = std::ratio_divide<P, nano_period>;
+
+    static_assert(
+      std::is_same_v<output_type, nano_rep>,
+      "Output type is not the same rep as std::chrono::nanoseconds");
+
+    // Extra check to ensure the output type is same as the underlying type
+    // supported in lib[std]c++.
+    static_assert(
+      std::is_same_v<
+        output_type,
+        int64_t> || std::is_same_v<output_type, long long>,
+      "Output type not in supported integer types.");
+
+    constexpr auto ratio_num = static_cast<output_type>(ratio::num);
+    const auto dur = static_cast<output_type>(duration.count());
+    output_type mul;
+    if (unlikely(__builtin_mul_overflow(dur, ratio_num, &mul))) {
+        // Clamp the value.
+        // Log here to ensure that it is picked up by duck tape. Ideally this
+        // should never happen, but in case it does, we have a log trail.
+        //
+        // If you are here because your ducktape test failed with this BadLogLine
+        // check, it means that you serialized a duration type that caused an
+        // overflow when casting to nanoseconds. Clamp your duration to
+        // nanoseconds::max().
+        using input_type = typename std::chrono::duration<R, P>;
+        vlog(
+          serde_log.error,
+          "Overflow or underflow detected when casting to nanoseconds, "
+          "clamping to a limit. Input: {}  type: {}",
+          duration.count(),
+          type_str<input_type>());
+        return duration.count() < 0 ? std::chrono::nanoseconds::min().count()
+                                    : std::chrono::nanoseconds::max().count();
+    }
+    // No overflow/underflow detected, we are safe to cast.
```

Comment:
> (avoiding repeated multiplication in duration_cast)

I thought of doing this but skipped it because it would be consistent  to rely on the output of duration_cast<> logic in both directions and just rely minimally on the builtin for overflow checks. Also because the logic is already complex here and it would only confuse the reviewers. But I agree that its a good idea.